### ISSUE TITLE
Deep integration with creme

### ISSUE THREAD

**Issue by MaxHalford** [vaexio/vaex#747]
I'm opening this issue to discuss some details that would allow the [creme library](https://github.com/creme-ml/creme) to make the most of vaex's engine. I haven't looked too deeply into this because until now we were not too concerned with performance.

We call "pure" online learning a scenario where a model learns with **one** observation at a time. We like this because it's elegant, easy on the implementation side, and simplifies things from a user perspective. However, we have to admit that the huge downside of our approach is that it's painfully slow. Essentially, training a creme model on a large dataset requires a Python for loop, which as we all know is slow.

To be able to do online learning in a fast way, we *have* to support mini-batches (e.g. the `partial_fit` from scikit-learn). The way we're thinking about this is that we don't want to change our design and philosophy. However, we're willing to add mini-batch support to a few selected models that 1) are performant 2) would benefit from vectorisation.

In terms of API, we're thinking of adding some `*_many` methods in addition to the `*_one` methods we already have. These methods would take as input a `pandas.DataFrame` and would output a `pandas.Series` or a `pandas.DataFrame`, depending on the context. We kind of have to with `pandas` because a big design choice of ours is to have access to feature names, so `numpy` arrays don't make the cut.

I sort of know how to proceed in terms of implementation. I know how to do this in a way that wouldn't require duplicating too much logic, and would provide the best of both worlds on our sides (pure online and mini-batching). As I've said to Jovan, I would rather fletch out with you guys what would be an ideal world on your side :). From what I understand that is being said in [this scikit-learn issue](https://github.com/scikit-learn/scikit-learn/pull/14963), you guys are essentially "intercepting" numpy ufunc calls and dispatching to whatever backend is available. Is that right? If so, I suppose that you would want to use numpy as much possible?

Also, if I understand correctly in your 1 billions rows in 20 minutes example, the `StandardScaler` class is benefiting from `vaex`, but not the `SGDClassifier` because it's implemented in Cython and not numpy? You'll have to excuse me if I'm completely missing the point, a lot of this is flying over my head :)

**Comment by maartenbreddels** [comment]
Dear Max,

@JovanVeljanoski had only nice things to say about creme, so we're happy you are reaching out to us.


> However, we're willing to add mini-batch support to a few selected models that 1) are performant 2) would benefit from vectorisation.

sounds like a good strategy.


> because a big design choice of ours is to have access to feature names, so `numpy` arrays don't make the cut.

This is one of the big reasons why vaex-ml and #415 exist, not losing your feature names, totally agree.

Regarding transfer of data, but being API neutral (or not really Pandas bound), you could either wait (there are standards in the making for DataFrames, but don't hold your breath). Or you could take a look at Apache Arrow. Vaex dataframe ingest and output Arrow Tables, and arrays, so that would be a perfect neutral transport mechanism.
With #415 it should also be possible to treat the incoming object as a (2d) numpy array, while it is actually a dataframe. You can call numpy function on it (those that are implemented), assign data to it, and create new 'arrays' and 'matrices/dataframes' using np.zeros_like etc.

Yes, the SGDClassifier is using sklearn 'normally', but while sklearn is fitting, vaex (in parallel) prepare the next batch, so sklearn never has to wait on vaex.

I hope this answers some of your questions, if not, feel free to ask more.

Regards,

Maarten

**Comment by JovanVeljanoski** [comment]
Hi Max, 

Also to clarify that in that article, we have re-implemented `StandardScaler` (it is part of vaex-ml), primarily for the reason of keeping track of feature names, and other convenience. Also, when it was 1st done NEP13/NEP18 did not exist. In the future we should to use sklearn transformers as much as possible.  If you look at this [sklearn PR](https://github.com/scikit-learn/scikit-learn/pull/14963) Maarten explains more about what we hope to achieve in the (near?) future. 

As for `SGDClassifier`, indeed we are just sending chunks of numpy data to it as it is, and sklearn does everything.

**Comment by MaxHalford** [comment]
Hey @maartenbreddels and @JovanVeljanoski, thanks for the detailed answers.

Things are still slightly fuzzy to me, but I think I'm getting there.

Say that we implement functions that accept a `pandas.DataFrame` for training and predicting, and output a `pandas.Series`/`pandas.DataFrame` for predicting. Would that work on your side? 

For instance, if we implement a `StandardScaler` that processes `pandas.DataFrames`, would that be good enough? If so, do we have to use `numpy` functions instead of `pandas` methods to benefit from `vaex`?

> As for SGDClassifier, indeed we are just sending chunks of numpy data to it as it is, and sklearn does everything.

Does this require making copies every time, or can there is some memory-view magic happening?

Kind regards :)

**Comment by MaxHalford** [comment]
So just to keep you updated, I've just written the mini-batch implementations for standard scaling and linear models. The relevant pull request is [here](https://github.com/creme-ml/creme/pull/299). The results are very good and the API is very practical, it's pandas in and pandas out. Essentially we can add/remove features or even change the order of columns and everything keeps working.

**Comment by JovanVeljanoski** [comment]
Actually.. i just realized perhaps there is absolutely no problem with using vaex with creme on arbitrarily large data, possibly without the need to modify either library at all :).

Vaex dataframes implement a `.to_pandas_df` method. But the method can also be used like this:

```
generator = df.to_pandas(chunk_size=10_000)   # or whatever size that optimizes model performance or speed
for i1, i2, chunk in generator:
    print(chunk)
```

so basically when the `chunk_size` argument is specified, the method returns a generator, which yields the start and end indices of the chunk, and a chunk of the specified size, which in this case is a pandas dataframe. 

So my idea is to use creme along these lines (pardon my lack of knowledge of the exact syntax):
```
generator = df.to_pandas(chunk_size=10_000)
for i1, i2, chunk in generator:
    creme_model.fit_batch(X=chunk[features], y=chunk[target])
```
which I was hoping it should be possible given the spirit of incremental learning (but now in batches). 

Would this help?

**Comment by MaxHalford** [comment]
Nice! I was kind of hoping that something like that was possible. However I suppose this would not make the most of the vaex engine, right? For instance it wouldn't allow making use of a GPU, or would it?

**Comment by JovanVeljanoski** [comment]
In the example above, vaex would be only used to create a chunk of data, in this case a pandas dataframe. Following that, anything done to that dataframe is out of the realm of vaex. 

However (if you are familiar with vaex a bit) we have a concept of "virtual column", which are lazily evaluated. So when one creates a pandas chunk, those virtual columns need to be materialized (i.e. put in memory), so acceleration via numba or cuda is possible for that step indeed. But what happens after that is up to creme, or whatever else users do. 

If you want vaex to do the computations, all the data manipulations done on the dataframe need to be via numpy, so then vaex can intercept them and do them, avoiding the need to convert from a vaex dataframe to a pandas dataframe or numpy arrays. This is exactly what @maartenbreddels tried to achieve with that sklearn PR. I hope this makes sense a bit?

**Comment by MaxHalford** [comment]
> I hope this makes sense a bit?

Yes it does! Thank you for being so patient with me.

> If you want vaex to do the computations, all the data manipulations done on the dataframe need to be via numpy, so then vaex can intercept them and do them, avoiding the need to convert from a vaex dataframe to a pandas dataframe or numpy arrays.

So the thing is that in fact we do everything via `numpy`. I found that calling `pandas` brings a lot of unnecessary overhead. We just use `pandas` for the feature names. Therefore I suppose we're good to go?

**Comment by JovanVeljanoski** [comment]
Yes well, if you are happy with my proposal with using `df.to_pandas_df(chunk_size=1000)` generator approach, than yeah. You are free to do as you please :). 

If you want to try and let vaex intercept numpy function calls, it would be better if @maartenbreddels takes a look at it to see if everything is indeed feasible (just in case). Maybe you can point out to us one of the algorithms so we can take a look? We've not implemented _all_ numpy methods, but I believe most, especially the common ones are in.

**Comment by MaxHalford** [comment]
Okay sounds good. The scaler code is [here](https://github.com/creme-ml/creme/blob/mini-batch/creme/preprocessing/scale.py#L74) (see `fit_many` and `transform_many`) whilst the linear mode code is [here](https://github.com/creme-ml/creme/blob/mini-batch/creme/linear_model/glm.py).
